2nd Draft of proposed Security Plugin API (CORE-207) (#68)
telicent-oss/smart-caches-core: docs/security/index.md

Author: Reviews should start by reading this documentation that aims to explain the API design in more technical detail and hopefully then makes the code portions of this PR more clearly understandable.

### `Entitlements<?>`

`Entitlements` is a generic type that holds user entitlements, it holds the `encoded()` byte sequence as a `byte[]`, a
`schema()` identifier as a `short`, and its `decodedEntitlements()` method provides access to the concrete entitlements
type a security plugin uses for its implementation.  Again the `AbstractSecurityPrimitive` class provides a basic
implementation of most of the interface.

### `SecurityPlugin`

`SecurityPlugin` is the top level entrypoint API by which applications interact with the Security Plugin API.  Primarily
this is just a place for applications to obtain instances of the other interfaces they need to carry out data processing
and access control tasks.

An implementation should generally make the instances of the other interfaces singletons where the API allows it e.g.
there should be a single shared `SecurityLabelsParser`.  However implementations **MUST** be aware that they will be
used in multi-threaded environments thus any singleton instance they return **MUST** be thread-safe.

Note that some of the interfaces a plugin provides instances of have specific lifecycles associated with them, e.g.

> `Authorizer`, so for those implementations **MUST** ensure they respect the defined lifecycle of that interface.

Author: - [ ] TODO: Consider making interfaces with short lifecycles clearer by making them `AutoCloseable` so the JVM will complain if applications don't consume and `close()` these appropriately.

This interfaces single `entitlementsForUser(Jws<Claims>, String)` method returns the generic `Entitlements<?>` type
allowing the plugin to obtain and supply entitlements however it sees fit.

The `EntitlementsProvider` to use is obtained via the `SecurityPlugin.entitlementsProvider()` method.

### `EntitlementsParser`

The `EntitlementsParser` is intended primarily as an internal implementation detail of a plugin, as for the most part
applications only need to work with `Entitlements<?>` as returned by the
[`EntitlementsProvider`](#entitlementsprovider).  However, in some cases an application may need to temporarily store
the entitlements, in which case they **MUST** store only the opaque byte sequence and use the plugins
`EntitlementsParser` to convert it back to a concrete `Entitlements<?>` object when it is needed again.

The `EntitlementsParser` to use is obtained via the `SecurityPlugin.entitlementsParser()` method.

### `SecurityLabelsParser`

The `SecurityLabelsParser` allows an application to take the label byte sequences it has encountered, or previously
stored, for data and turn them back into `SecurityLabels<?>` that can be used with an [`Authorizer`](#authorizer) to

> make access decisions.  It's single `parseSecurityLabels(byte[])` method takes a raw byte sequence and produces a

Author: ```suggestion
make access decisions.  It has a single `parseSecurityLabels(byte[])` method takes a raw byte sequence and produces a
```

when processing a single RDF event from the `knowledge` topic, thus applications obtain instances of these by calling
the `SecurityPlugin.prepareLabelsApplicator(byte[], Graph)` method providing the labels byte sequence from the
`Security-Label` header and the fine-grained labels graph (if any).  Therefore plugins **MUST** provide a fresh instance
of this interface every time their `prepareLabelsApplicator()` method is invoked as the labels that apply **MAY** change
on every RDF event an application processes.

Once an application has an instance they call `labelForTriple(Triple)` whenever they need to lookup the label for a
particular triple, this returns a `SecurityLabels<?>` from which the application can then store the corresponding byte
sequence by accessing its `encoded()` method.

### `Authorizer`

The `Authorizer` allows an application to make access decisions based on the labels for data.  This is intended to be
scoped to the lifetime of a single user request so an instance is created by calling the
`SecurityPlugin.prepareAuthorizer(Entitlements<?>)` method passing in the users entitlements (see [Example
Usage](#example-usage)) for a fuller outline usage.

> 

Author: - [ ] Incomplete description here
- [ ] Make reference to caching for performance again

## Example Usage

In the following example we show the complete lifecycle of how an application would make an access decision:

```java
// Load the plugin, this is cached for the lifetime of the JVM so repeated calls are cheap
SecurityPlugin plugin = SecurityPluginLoader.load();

// Get the Users Entitlements and prepare an Authorizer based upon those
String userId = plugin.identityProvider().identityForUser(jws);
Entitlements<?> entitlements = plugin.entitlementsProvider().entitlementsForUser(userId);

> Authorizer authorizer = plugin.prepareAuthorizer(entitlements);

Author: - [ ] Use in `try-with-resources` block

// Get the Labels Parser
SecurityLabelsParser parser = plugin.labelsParser();

// Get results from the underlying data store
List<SomeResult> results = runQuery();
for (int i = 0; i < results.size(); i++) {
  // Filter the results based on the labels present
  byte[] rawLabels = results.get(i).getSecurityLabels();
  try {
    SecurityLabels<?> labels = parser.parseSecurityLabels(rawLabels);
    if (!authorizer.canAccess(labels)) {
        results.removeAt(i);
        i--;
    }
  } catch (MalformedLabelsException e) {
    // Could be thrown if the labels are not supported by this plugin, or otherwise malformed
    results.removeAt(i);
    i--;
  }
}

// Return the filtered results
return results;
```

Firstly an application has to obtain the plugin instance, most applications will probably do this once early in their
startup and make the instance accessible wherever it is needed.  It then uses other parts of the plugin API to determine
the user identity, obtain their entitlements and prepare an authorizer based upon those.  Once it has an authorizer it
can then make any access decisions necessary to fulfill the current request/operation before producing the filtered

> results.

Author: - [ ] Add clarifying note about `Authorizer` lifecycle here and recommend `try-with-resources` pattern

  register the plugin.

By implementing things in this way existing implementations of different label schemas can be combined by creating a new
plugin that depends on the logic modules of one/more existing plugin implementations, with its implementations wrapping
those existing implementations.  This can then be registered as a combined plugin via its own registration module.

### Caching for Performance

In real deployments the actual pool of distinct labels is often relatively small.  Therefore implementations can
generally make intelligent caching decisions, e.g. caching the results of parsing label byte sequences into
implementation specific data structures.  Over the lifetime of an application which may make hundreds of
thousands/millions of access decisions this can offer significant performance benefits.

Similarly within an `Authorizer` instance an implementation may wish to cache access decisions so frequently seen labels
can be rapidly evaluated.

By placing such concerns onto plugin implementations we aim to allow applications to remain unaware of this concern and
avoid littering each application with their own caching logic.

### Plugin Dependencies

> A Security Plugin **SHOULD** depend on `security-core` as de

Author: - [ ] Unfinished section

# Security Plugins

The `SecurityPlugin` interface provides an abstraction around the Telicent Core Platform's label based security model
allowing applications to enforce the model without needing any direct knowledge of how the labels are interpreted and
enforced.

## Design

In this section we first outline the design motivation behind the interfaces that are defined, see
[Interfaces](#interfaces) for outlines of the various interfaces and their usages.  As with everything else in our Core
Libraries this follows our general [Design Ethos](../design.md#design-ethos).

### Label Format & Syntax

As background, remember that all events flowing through Kafka within the Core Platform are expected to have a

> `Security-Label` header present that indicates the labels that apply to the data.  Additionally, RDF format events on

Reviewer: What's the future for the `policyInformation` header? That currently contains either an EDH or IDH where the access control parts are converted into a security label. That was a stop-gap to allow us to develop federation and explore some of the principles you've laid out in this PR. In practice it works but is confusing for customers.

Reviewer: I'd like to have a crack at implementing these interfaces with the scheme we used in the federation filter.

Author: So ideally the EDH/IDH is just put directly in the `Security-Label` header and you deploy a plugin that knows how to enforce the access criteria that the policy defines.

Yes our goal is to have probably 2 supported plugins - RDF-ABAC and EDH/IDH - which we build out as we start adopting this API and adapting it to address the inevitable implementation issues we discover

Author: Added some clarifications around this

the `knowledge` topic *MAY* optionally include a labels graph that provides fine grained labels for individual triple(s)
within the event.  Currently these labels are expressed as [RDF-ABAC Attribute Expressions][RdfAbac] which are strings
declaring an Attribute Based Access Control (ABAC) policy in the form of attribute expressions.

As part of evolving the platform towards a more flexible Policy Based Access Control (PBAC) model the new plugin APIs
introduced from `0.25.0` onwards instead treat the `Security-Label` as opaque byte sequences.  These byte sequences have
an optional 4 byte schema prefix to allow plugins to detect whether they support a particular label schema and reject
labels they don't support.  This also potentially provides the capability for future plugins to provide support for

> multiple label schemas coexisting within the platform.

Reviewer: Why? Managing a single ABAC scheme consistently at an enterprise level is hard enough especially so for multiple co-operating organisations.

Author: It's more a theoretical capability for future proofing than anything else.  I don't anticipate us actually supporting this anytime soon.

Also having an explicit schema prefix makes it easier to evolve labelling schemas that don't naturally have their own versioning mechanism built-in e.g. current usage of string based labels.

Author: Added clarifications that while this is theoretically supported we have **NO** intention of doing this anytime soon

[Apache Avro][Avro], [Apache Thrift][Thrift] etc.

The `SecurityLabels<T>` interface treats labels in abstract terms, allowing API users to be treat the labels in entirely
abstract terms, while a plugin can internally decode the encoded byte sequence into whatever data structures it needs to
operate over.  A plugins `SecurityLabelsParser` converts from the opaque byte sequences into a plugins specific
`SecurityLabels` implementation, note that there is no corresponding encoding interface needed as access to the opaque
byte sequence is always provided by the `SecurityLabels` interface so applications that need to store labels for later
evaluation can simply store the opaque byte sequence directly.

#### Notes on encoding fine-grained labels

This change does not preclude the existing usage of fine-grained label graphs, but again some evolution is needed.  A
labels graph means we are using a concrete RDF syntax, e.g. Turtle, so we are constrained to encoding labels as strings
in those syntaxes.  Therefore, in order to support arbitrary label schemas as described above the labels graph **MUST**

> encode labels as [`xsd:base64Binary`](XsdBase64) typed literals e.g.

Reviewer 2: Link needs to be `[XsdBase64]`.

Author: ```suggestion
encode labels as [`xsd:base64Binary`][XsdBase64] typed literals e.g.
```

left to plugins via their `EntitlementsProvider` and `EntitlementsParser` implementations.  This means that the burden
of configuring any necessary supporting services, e.g. connectivity to Access, becomes an internal implementation detail
of a plugin, rather than an application concern. Entitlements use the same optional schema prefix as part of the byte
sequences to allow plugins to distinguish between different entitlement schemas they may support, and this allows for
future plugins to potentially combine multiple entitlement providers.

As with labels the `Entitlements<T>` interface treats entitlements in abstract terms from an application perspective
while allowing a plugin implementation to decode the entitlements into whatever data structure(s) are appropriate. A
plugins implementation of the `EntitlementsParser` interface can supply whatever decoding logic it needs for its
implementation.

### Data Access Enforcement

Data Access Enforcement is done by combining the users retrieved [Entitlements](#entitlements) with the
[Labels](#label-format--syntax) present on the data.  This is done by first preparing an `Authorizer` via the

> `prepareAuthorizer()` method passing in the users entitlements.  `Authorizer` instances are intended to be scoped to the

Reviewer: Could this be expanded to include some flexible context as well as the user's entitlement?

For example ODY have suggested policies/conditions based on location and device authentication. These attributes will likely need to come from the connection / request somehow.

Author: The thinking, which probably wasn't clear enough here, is that the entitlements information that gets retrieved would include that kind of detail.

Of course it may be that the interfaces as defined in this 1st draft aren't sufficient for that, e.g. perhaps `prepareAuthorizer()` needs to take an additional parameter which provides access to the connection/request details somehow

lifetime of a single user request so they may cache any access decisions if they encounter the same labels repeatedly.

> Once an application has an `Authorizer` it calls the `canAccess()` method passing in the labels for the data it needs an

Reviewer: Could this be expanded to support other decision types? Currently `canAccess()` limits the decision to reading of the data, however, we know there are requirements for writing data, updating data, updating security labels etc.

Further, we know that we need to apply robust access control at the business logic level which has currently has a big sticking plaster over it using Istio at the API request level. Could this be used to solve that too?

Author: Almost certainly so, again my main focus for the 1st draft was defining something that is functionally equivalent to what we currently do with RDF-ABAC so we can wrap it up and validate the proposed API.

What's here is definitely a subset of what was in the larger design document where we were discussing using this for other access decisions like writes, maybe at a minimum this part of the API should be:

- `canRead()` for data reads
- `canWrite()` for data writes
- `canAccess()` for API routes/business logic

The last one probably needs more design thinking as we have to consider how the API routes/business logic actions get labelled, and how that would interact with allowing for arbitrary labelling schemas.

Author: API has been evolved along the lines suggested above in 2nd Draft

access decision for.  This returns either `true` for accessible, or `false` for forbidden.  In the event of any
problem/ambiguity in computing an access decisions plugins **MUST** fail-safe by defaulting to returning `false` if they
can't make an access decision.

You can see the complete lifecycle outlined later in [Example Usage](#example-usage).

### Plugin Loading

Plugins are loaded by applications by calling `SecurityPluginLoader.load()` which uses the JVM `ServiceLoader` to locate
registered plugins.  The loaded plugin is cached for the lifetime of the JVM so subsequent calls just return the
previously loaded instance.

The expectation is that there is `1`, and **ONLY** `1` plugin registered in this way.  If no plugin is registered, more
than 1 plugin is registered, or loading the registered plugin throws an error, then the system defaults to a fail-safe
mode and instead loads the unregistered `FailSafePlugin`. In fail-safe mode the system treats all labels and

> entitlements as invalid and defaults all access decisions to forbidden.  This will be clearly and explicitly noted in

Author: > Given how vital this is should we force the app to shutdown if there is an issue?

Throwing an error is our best effort to do this.  We could `System.exit()` but that tends to make writing tests around the behaviour very difficult.

While we control our applications customers can build their own applications that integrate our security model using this API and we won't be able to control how they call the API.  So falling back into a fail-safe mode is a compromise solution to try and protect developers from themselves.

> entitlements as invalid and defaults all access decisions to forbidden.  This will be clearly and explicitly noted in

Author: > What happens if we can't connect to Kafka or Access? I guess we should be consistent

For Kafka it depends on the application, generally it will just keep trying to connect and many eventually time out and crash restart.

For Access the application tends to get an error retrieving user attributes, again depending on the application this either manifests as a user visible error or their request is treated as having no attributes so they get minimal access to data.

### `SecurityPlugin`

`SecurityPlugin` is the top level entrypoint API by which applications interact with the Security Plugin API.  Primarily
this is just a place for applications to obtain instances of the other interfaces they need to carry out data processing
and access control tasks.

An implementation should generally make the instances of the other interfaces singletons where the API allows it e.g.
there should be a single shared `SecurityLabelsParser`.  However implementations **MUST** be aware that they will be
used in multi-threaded environments thus any singleton instance they return **MUST** be thread-safe.

Note that some of the interfaces a plugin provides instances of have specific lifecycles associated with them, e.g.
`Authorizer`, so for those an implementation **MUST** ensure they respect the defined lifecycle of that interface.
These interfaces are clearly marked by having them extend `AutoCloseable` so that the Java compiler will issue warnings
if application developers are not actively `close()`'ing the obtained instances, or using them in a `try-with-resources`
block to implicitly `close()` them as shown in the [Example Usage](#example-usage).

> ### `IdentityProvider`

Author: To my mind yes, as you comment elsewhere, maybe in conjunction with the Request/Connection details to give more context for access decisions.

For the delegated access stuff we probably want to consider leveraging more aspects of the OAuth 2/OIDC protocols to allow users to authorize specific applications to act as them at which point those applications would get Access and Refresh tokens for the users (in the form of JWTs).  Obviously here be dragons around secure storage of said credentials once obtained (but can hopefully be delegated to secure secret storage in the deployment)!

The Core Platform relies upon Open ID Connect (OIDC) as the authentication mechanism and uses Bearer tokens in the form
of cryptographically signed JSON Web Tokens (JWTs).  Depending on how a given deployment is configured there *MAY* be
multiple underlying identity providers, but by design individual applications don't need to be aware of these, they
merely need to be able to cryptographically verify the presented tokens using the public keys of the identity providers.

However, in practical terms most applications deal with users, rather than tokens, when making access decisions.  Thus a
key part of a plugin is the `IdentityProvider`.  This interface translates from the authenticated users JWT into a user
identifier that the application can use, typically this is some form of username/email address.  However, depending on
the security plugin and/or deployment this may need to be a more opaque identifier since email addresses may be reused
across multiple identity providers e.g. some combination of the `iss` and `sub` claims from the JWT.

For most plugins we don't expect the mapping of JWT to user identity to vary so we provide a generic

> `DefaultIdentityProvider` that we expect most plugins will reuse.  This maps the user identity by looking for claims in

Reviewer: Suggest the claim used as the principal's identifier should be fixed in configuration rather than fall back through a list of preferences.

Author: Clarified that this is fully configurable, and added a note below stating that a customer could fix this to a specific claim if they so desired.

implementations, of the API to provoke specific error handling conditions/scenarios that are necessary to achieve
appropriate test coverage.

## Implementation Concerns

### Separating Logic and Registration

As described [earlier](#plugin-loading) the system expects `1`, and only `1`, plugin to be registered.  At the same time
it is also designed to allow for multiple label and entitlement schemas to potentially co-exist in a single plugin.

This means that generally implementations should be done as two modules:

- A Logic module that provides the actual interface implementations that implement a particular security model/policy
  framework within the API.
- A Registration module that depends on the logic module and provides the appropriate `META-INF/services` file to
  register the plugin.

By implementing things in this way existing implementations of different label schemas can be combined by creating a new
plugin that depends on the logic modules of one/more existing plugin implementations, with its implementations wrapping
those existing implementations.  This can then be registered as a combined plugin via its own registration module.

### Caching for Performance

> In real deployments the actual pool of distinct labels is often relatively small.  Therefore implementations can

Reviewer: Is this true? We don't have any fully operational customer deployments yet. The features ODY are putting forward imply this won't necessarily be the case.

Author: I think it's only really true for our internal development clusters, for actual customer deployments there seems to be far larger variation in labelling so need to rephrase this section to make it clearer

Author: Section has been reworded accordingly

transitive dependencies from this) as the applications will already have their own dependency on this.

Plugins **MAY** need to have their own dependencies, e.g. a particular security/policy language library, and whatever
other supporting dependencies it needs.  However plugin authors **MUST** be aware that as described under [Plugin
Loading](#plugin-loading-in-deployments) a plugin is always added to the end of the classpath.  This means that plugin
authors need to be careful that they aren't relying on any dependencies that have a version mismatch with those being
used in applications otherwise various JVM Classloading errors may occur at runtime are prevent correct operation of a
plugin.

Generally plugin authors should use the managed dependencies in this repository as a guide to what common dependencies
are likely be to be encountered across applications and try to reuse those wherever appropriate, e.g. using Jackson for
JSON processing.

Ideally plugin dependencies **SHOULD** be as minimal as possible, and when packaged into an image for deployment
**SHOULD** exclude any dependency that they can safely assume will already be present e.g. `security-core`.

> ### Impacts on Existing Code

Author: ```suggestion
### Impacts on Existing Code
```
